src/Components/test/E2ETest/Tests/DownloadAnchorTest.cs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using BasicTestApp;
+using BasicTestApp.RouterTest;
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure;
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure.ServerFixtures;
+using Microsoft.AspNetCore.Components.E2ETest.ServerExecutionTests;
+using Microsoft.AspNetCore.E2ETesting;
+using Microsoft.Extensions.DependencyInjection;
+using OpenQA.Selenium;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.AspNetCore.Components.E2ETest.Tests
+{
+    public class DownloadAnchorTest
+        : ServerTestBase<ToggleExecutionModeServerFixture<Program>>
+    {
+      
+
+        public DownloadAnchorTest(
+            BrowserFixture browserFixture,
+            ToggleExecutionModeServerFixture<Program> serverFixture,
+            ITestOutputHelper output)
+            : base(browserFixture, serverFixture.WithServerExecution<Program, TestServer.HostedInAspNetStartup>(), output)
+        {            
+        }
+
+        [Fact]
+        public void DownloadFileFromAnchor()
+        {
+            // Arrange
+            MountAndNavigateToRouterTest();
+            var initialUrl = Browser.Url;
+
+            // Act
+            var anchor = Browser.Exists(By.CssSelector("a[download]"));
+            anchor.Click();
+
+            // URL should still be same as before click
+            Assert.Equal(initialUrl, Browser.Url);
+
+            // File should be requested
+            var requestedPaths = GetAndClearRequestedPaths();
+            Assert.NotEmpty(requestedPaths.Where(path => path.EndsWith("Download.txt", StringComparison.InvariantCultureIgnoreCase)));
+        }
+
+        protected IWebElement MountAndNavigateToRouterTest()
+        {
+            Navigate(ServerPathBase, noReload: true);
+            var appElement = Browser.MountTestComponent<TestRouter>();
+            return appElement;
+        }
+
+
+        private IReadOnlyCollection<string> GetAndClearRequestedPaths()
+        {
+            var requestLog = _serverFixture.Host.Services.GetRequiredService<TestServer.ResourceRequestLog>();
```

Comment:
@javiercn how do you feel about this?

Reply:
@mrdav1111 There's a way to validate what resources were requested in the test harness without needing an additional utility.

You can check out how it is done in the lazy-loading tests.

https://github.com/dotnet/aspnetcore/blob/8dedfde5c0f3c1c16185efed5912469a5f565f82/src/Components/test/E2ETest/Tests/WebAssemblyLazyLoadTest.cs#L169-L179

The feature works by leveraging Chrome's diagnostics API to get the list of requested resources.

Reply:
@captainsafia I tried this method but seems like `window.performance.getEntriesByType('resource')` just log the resources requested by the application (DOM, JavaScript) and not the file downloads.

I took that logic from the `BootResourceCachingTest` as mentioned on https://github.com/dotnet/aspnetcore/issues/20360#issuecomment-690353608
https://github.com/dotnet/aspnetcore/blob/27341f9be1a3cd9fac8eef6cccef1dc80b6af2e0/src/Components/test/E2ETest/Tests/BootResourceCachingTest.cs#L140-L148

Reply:
Is it important to verify that the file is downloaded?

Reply:
I see. I hadn't realized that we already use this pattern elsewhere.

> Is it important to verify that the file is downloaded?

It's not important that we validate that the the file was downloaded but IMO it doesn't hurt to have this check.

Reply:
> Is it important to verify that the file is downloaded? 

@pranavkm I agree that it is not strictly necessary for this bug.
>EDIT: Actually, after reading the changes that were made in the ServerStartup, I'd favor the approach of using the `window.performance` API. This way, we avoid having test case specific logic in the test server's startup code.

@captainsafia I already tried to use the `window.performance` API but `window.performance.getEntriesByType('resource')` just contain requests initiated by the DOM or JavaScript loading the application resources, this API does not measure file downloads out of that. So the `HasLoadedAssembly` method always return false even if the file has been downloaded.

I can't come up with a solution for this part of the test just using Selenium or JavaScript.

Should the file download assertion be removed?

Reply:
We discussed this during our engineering sync.

For now, using this pattern (request log and middleware) is fine for a single test.

However, we should be careful to not overuse this pattern in the future so we don't end up having a middleware for each individual test case.

Thanks for taking a look into this!